File: Ejercicios/T7/PYTHON/detector_peons.py
import os
import glob
import cv2
import numpy as np

# Inicialozamos parametros HoG
winSize = (64, 128)
blockSize = (16, 16)
blockStride = (8, 8)
cellSize = (8, 8)
nbins = 9
derivAperture = 1
winSigma = -1
histogramNormType = cv2.HOGDESCRIPTOR_L2HYS
L2HysThreshold = 0.2
gammaCorrection = True
nlevels = 64
signedGradient = False

# HOG
hog = cv2.HOGDescriptor(winSize, blockSize, blockStride,
                        cellSize, nbins, derivAperture,
                        winSigma, histogramNormType, L2HysThreshold,
                        gammaCorrection, nlevels, signedGradient)

# cargamos o modelo entrenado
model = cv2.ml.SVM_load('../data/models/pedestrian.yml')

sv = model.getSupportVectors()
rho, aplha, svidx = model.getDecisionFunction(0)
svmDetectorTrained = np.zeros(sv.shape[1] + 1, dtype=sv.dtype)
svmDetectorTrained[:-1] = -sv[:]
svmDetectorTrained[-1] = rho

# Instanciamos o SVMDetector entrenado por nos en HOG
hog.setSVMDetector(svmDetectorTrained)

# HOG de Opencv para detectar peons
hogDefault = cv2.HOGDescriptor(winSize, blockSize, blockStride,
                               cellSize, nbins, derivAperture,
                               winSigma, histogramNormType, L2HysThreshold,
                               gammaCorrection, nlevels, signedGradient)
svmDetectorDefault = cv2.HOGDescriptor_getDefaultPeopleDetector()
hogDefault.setSVMDetector(svmDetectorDefault)

# Lemos as imaxes dos peons dende disco
imageDir = '../data/images/pedestrians'
imagePaths = glob.glob(os.path.join(imageDir, '*.jpg'))

# Normalizamos todas as imaxes a mesma altura
finalHeight = 800.0

for imagePath in imagePaths:
  print('procesando: {}'.format(imagePath))

  # Lemos a imaxe
  im = cv2.imread(imagePath, cv2.IMREAD_COLOR)

  # redimensionamos a imaxe a altura de finalHeight
  scale = finalHeight / im.shape[0]
  im = cv2.resize(im, None, fx=scale, fy=scale)

  # detectMultiScale co detector entrenado por nos
  bboxes, weights = hog.detectMultiScale(im, winStride=(8, 8), padding=(32, 32),
                                         scale=1.05, finalThreshold=2,
                                         hitThreshold=1.0)

  # detectMultiScale empregando o detector por defecto de OpenCV
  bboxes2, weights2 = hogDefault.detectMultiScale(im, winStride=(8, 8), padding=(32, 32),
                                                  scale=1.05, finalThreshold=2,
                                                  hitThreshold=0)

  # Imprimimos os peons detectados
  if len(bboxes) > 0:
    print('Detector entrenado :: peon detectado: {}'.format(bboxes.shape[0]))
  if len(bboxes2) > 0:
    print('Detector OpenCV :: peon detectado: {}'.format(bboxes2.shape[0]))

  # Debuxamos os rectangulos contedores
  # Roxo = detector OpenCV, verde = Detector entrenado
  for bbox in bboxes:
    x1, y1, w, h = bbox
    x2, y2 = x1 + w, y1 + h
    cv2.rectangle(im, (x1, y1), (x2, y2), (0, 255, 0), thickness=3, lineType=cv2.LINE_AA)

  for bbox in bboxes2:
    x1, y1, w, h = bbox
    x2, y2 = x1 + w, y1 + h
    cv2.rectangle(im, (x1, y1), (x2, y2), (0, 0, 255), thickness=3, lineType=cv2.LINE_AA)

  # Visualozamos o resultado final
  cv2.imshow("imaxe",im)
  cv2.waitKey(0)
  # Slavamos a disco
  imResultPath = os.path.join('results', os.path.basename(imagePath))
  cv2.imwrite(imResultPath, im)

File: Ejercicios/T7/C++/detector_peons.cpp
#include <iostream>
#include <dirent.h>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include "opencv2/objdetect.hpp"
#include <opencv2/ml.hpp>

using namespace cv::ml;
using namespace cv;
using namespace std;

void getFileNames(string dirName, vector<string> &imageFnames, string imgExt)
{
  DIR *dir;
  struct dirent *ent;
  int count = 0;

  vector<string> files;
  if ((dir = opendir (dirName.c_str())) != NULL) {
    while ((ent = readdir (dir)) != NULL) {
    // descartamos nomes de ficheiros non desexador
    if(strcmp(ent->d_name,".") == 0 | strcmp(ent->d_name, "..") == 0) { continue; }
    string temp_name = ent->d_name;
    files.push_back(temp_name);
    }

    // Ordenamos
    std::sort(files.begin(),files.end());
    for(int it=0;it<files.size();it++) {
      string path = dirName;
      string fname = files[it];
      if (fname.find(imgExt, (fname.length() - imgExt.length())) != std::string::npos) {
        path.append(fname);
        // imageFnames.push_back(path);
        imageFnames.push_back(fname);
      }
    }
    closedir (dir);
  }
}

int main() {
  //  HOG
  HOGDescriptor hog(
    Size(64, 128), //winSize
    Size(16, 16),  //blocksize
    Size(8, 8),    //blockStride,
    Size(8, 8),    //cellSize,
        9,     //nbins,
        0,     //derivAperture,
        -1,    //winSigma,
        HOGDescriptor::HistogramNormType (0),     //histogramNormType,
        0.2,   //L2HysThresh,
        1,     //gammal correction,
        64,    //nlevels=64
        0);    //signedGradient

  Ptr<SVM> svm = ml::SVM::load("../../data/models/pedestrian.yml");
  // vectores soporte
  Mat sv = svm->getSupportVectors();
  // funcion de decision
  Mat alpha, svidx;
  double rho = svm->getDecisionFunction(0, alpha, svidx);

  vector<float> svmDetectorTrained;
  svmDetectorTrained.clear();
  svmDetectorTrained.resize(sv.cols + 1);
  for (int j = 0; j < sv.cols; j++) {
    svmDetectorTrained[j] = -sv.at<float>(0, j);
  }
  svmDetectorTrained[sv.cols] = (float)rho;

  // SVMDetector entrenado por nos con HOG
  hog.setSVMDetector(svmDetectorTrained);

  // HOG detector de OpenCv que detecta persoas
  HOGDescriptor hogDefault(
    Size(64, 128), //winSize
    Size(16, 16),  //blocksize
    Size(8, 8),    //blockStride,
    Size(8, 8),    //cellSize,
        9,     //nbins,
        0,     //derivAperture,
        -1,    //winSigma,
        HOGDescriptor::HistogramNormType (0),     //histogramNormType,
        0.2,   //L2HysThresh,
        1,     //gammal correction,
        64,    //nlevels=64
        0);    //signedGradient
  // Inicializamos o detector.
  vector< float > svmDetectorDefault = hog.getDefaultPeopleDetector();
  hogDefault.setSVMDetector(svmDetectorDefault);

  string imageDir = "../../data/images/pedestrians/";
  vector<string> imagePaths;
  getFileNames(imageDir, imagePaths, "jpg");

  float finalHeight = 800.0;
  for (int i = 0; i < imagePaths.size(); i++) {
    cout << "procesando: " << imagePaths[i] << endl;
    Mat im = imread(imageDir + imagePaths[i], IMREAD_COLOR);

    float finalWidth = (finalHeight * im.cols) / im.rows;
    resize(im, im, Size(finalWidth, finalHeight));

    vector<Rect> bboxes, bboxes2;
    vector<double> weights, weights2;

    float hitThreshold = 1.0;
    Size winStride = Size(8, 8);
    Size padding = Size(32, 32);
    float scale = 1.05;
    float finalThreshold = 2;
    bool useMeanshiftGrouping = 0;
    hog.detectMultiScale(im, bboxes, weights, hitThreshold, winStride, padding,
               scale, finalThreshold, useMeanshiftGrouping);

    hogDefault.detectMultiScale(im, bboxes2, weights2, 0, winStride, padding,
               scale, finalThreshold, useMeanshiftGrouping);

    if (!bboxes.empty()) {
      cout << "Detector entrenado :: peons detectados: " << bboxes.size() << endl;
      vector< Rect >::const_iterator loc = bboxes.begin();
      vector< Rect >::const_iterator end = bboxes.end();
      for( ; loc != end ; ++loc ) {
        rectangle(im, *loc, Scalar(0, 255, 0), 2);
      }
    }
    if (!bboxes2.empty()) {
      cout << "Detector OpenCV :: peons detectados:: " << bboxes2.size() << endl;
      vector< Rect >::const_iterator loc = bboxes2.begin();
      vector< Rect >::const_iterator end = bboxes2.end();
      for( ; loc != end ; ++loc ) {
        rectangle(im, *loc, Scalar(0, 0, 255), 2);
      }
    }

    imshow("peons", im);
    imwrite("../results/" + imagePaths[i], im);
    waitKey(0);
  }
  return 0;
}
